### PR TITLE
esys: add SM4 algorithm support.

### PR DESCRIPTION
**Pull request by mayuanchenma**
add sm4 algorithm support.

### CODE REVIEW

File: src/tss2-esys/esys_crypto_ossl.h [hunk 1]
```diff
@@ -109,6 +109,28 @@ TSS2_RC iesys_cryptossl_sym_aes_decrypt(
     uint8_t *iv,
     void *userdata);
 
+#if HAVE_EVP_SM4_CFB && !defined(OPENSSL_NO_SM4)
```
Comment (williamcroberts) [inline src/tss2-esys/esys_crypto_ossl.h:112]: Do we really need the check with OPENSSL_NO_SM4, wont, the check to determine HAVE_EVP_SM4_CFB be undefined if OpenSSL is built without SM4?

File: src/tss2-esys/esys_crypto_ossl.c [hunk 2]
```diff
@@ -1212,6 +1212,164 @@ iesys_cryptossl_sym_aes_decrypt(uint8_t * key,
...
+
+    if (tpm_sym_alg != TPM2_ALG_SM4) {
+        goto_error(r, TSS2_ESYS_RC_BAD_VALUE,
+                   "SM4 decrypt called with wrong algorithm.", cleanup);
+    }
+
+    if (key_bits == 128 && tpm_mode == TPM2_ALG_CFB)
+        cipher_alg = EVP_sm4_cfb128();
+    else {
+
```
Comment (williamcroberts) [inline src/tss2-esys/esys_crypto_ossl.c:1339]: extra newline

**Review by williamcroberts — CHANGES_REQUESTED** [review]
Overall looks good, i'd like to see a test that includes an explicit usageof SM4 and skipped if the TPM doesn't support it. We should probably look at building swtpm on the test images with SM4 support as well.

File: src/tss2-esys/esys_crypto_ossl.h (hunk 1 above)
Comment (mayuanchenma) [reply src/tss2-esys/esys_crypto_ossl.h:112]: the same as SM3 check. @williamcroberts

**Comment by mayuanchenma** [comment]
@williamcroberts

File: src/tss2-esys/esys_iutil.c [hunk 3]
```diff
@@ -866,7 +928,7 @@ iesys_decrypt_param(ESYS_CONTEXT * esys_context)
         return_if_error(r, "Setting plaintext");
     } else {
         return_error(TSS2_ESYS_RC_BAD_VALUE,
-                     "Invalid symmetric algorithm (should be XOR or AES)");
+                     "Invalid symmetric algorithm (should be XOR or AES or SM4)");
```
Comment (williamcroberts) [inline src/tss2-esys/esys_iutil.c:931]: Change the wording here to:
"Should be XOR, AES, or SM4"

File: src/tss2-esys/esys_crypto_ossl.c (hunk 2 above)
Comment (williamcroberts) [reply src/tss2-esys/esys_crypto_ossl.c:1340]: Still need to remove this newline

**Review by williamcroberts — CHANGES_REQUESTED** [review]

**Comment by williamcroberts** [comment]
> @williamcroberts

Fix those issues for the comments I left and assuming the CI is happy we can pull it in after 4.0.0 release so it will be for 4.1.0+

**Comment by williamcroberts** [comment]
I think you'll also want to rebase this ontop of current master:
```bash
git pull --rebase origin master
```

**Review by williamcroberts — APPROVED** [review]

**Pull request merged**